miio/gateway.py

```diff
@@ -529,45 +580,261 @@ def set_light(self, color_name, brightness):
         brightness_and_color = brightness_and_color_to_int(
             brightness, color_map[color_name]
         )
-        return self.send("set_rgb", [brightness_and_color])
+        return self._gateway.send("set_rgb", [brightness_and_color])
 
 
-class SubDevice:
-    def __init__(self, gw, sid, type_, _, __, ___):
-        self.gw = gw
-        self.sid = sid
-        self.type = DeviceType(type_)
+class SubDevice():
+    """
+    Base class for all subdevices of the gateway
+    these devices are connected through zigbee.
+    """
 
-    def unpair(self):
-        return self.gw.send("remove_device", [self.sid])
+    def __init__(
+        self,
+        gw: Gateway = None,
+        dev_info: SubDeviceInfo = None,
+    ) -> None:
+        self._gw = gw
+        self.sid = dev_info.sid
+        self._battery = None
+        self._fw_ver = dev_info.fw_ver
+        try:
+            self.type = DeviceType(dev_info.type_id)
+        except ValueError:
+            self.type = DeviceType(-1)
+        
+        # if dataclass props is defined, initialize an instance
+        if hasattr(self, 'props'):
+            self._props = self.props()
```

Comment:
One more comment for this. You should simply initialize it always (and add an empty props for the main class), this way there is no need to do any sort of checking and the API is the same no matter what sort of device it is and what it exposes.

Reply:
done